### ISSUE TITLE
[RFC] Pluggable styling API

### ISSUE THREAD

**Issue by kettanaito** [kettanaito/atomic-layout#57]
## What

I propose to establishing a support for different CSS-in-JS libraries to help developers take benefits of Atomic Layout using their favorite styling solution.

## Why

* The library operates on CSS properties Object, which can be computed into string (when needed), and be compliant with about any modern CiJ solution

## Implementation

To enable such support it has been decided to convert the library to a monorepo that consist of the following parts:

- core library (`@atomic-layout/core`, responsible for media queries calculation, areas parsing, parametric components generation without attachment to any specific styling/rendering solution)
- rendering libraries (i.e. `@atomic-layout/emotion`)

All rendering libraries would list `@atomic-layout/core` as a dependency and utilize functions and types the core library exports.

## Usage

To use Atomic Layout with another styling solution (implying that such solution is supported) import the respective `@atomic-layout/X` package and use it with the same API described in the documentation:

```jsx
import React from 'react'
import { Composition } from '@atomic-layout/emotion'

const MyComponent = () => (
  <Composition templateCols="250px 1fr">
    <span>Emotion support</span>
    <span>Example</span>
  </Composition>
)
```

**Comment by kettanaito** [comment]
This most likely means that this repository needs to be published as a mono-repo.

```
src/
  packages/
    react/
      package.json
    vue/
      package.json
    .../
  shared/
    shared-logic-here
```

Sources are built into the root directory grouped by the domain name (i.e. `./vue`). Default export is still the React version.

**Comment by bitttttten** [comment]
Following from #122 I can think of 2 options where you can let the developer configure how styles and class names are generated (a), or configure which library is rendering the components (b).

#122 contains examples of option a.

Since all atomic-layout really needs to do is attach classnames and is ideally framework/component agnostic, option b should be possible.

Just to generate some conversation, I will provide this example showcasing both:

```diff
diff --git a/src/components/Composition.tsx b/src/components/Composition.tsx
index c9905d3..dd9c7bf 100644
--- a/src/components/Composition.tsx
+++ b/src/components/Composition.tsx
-const CompositionWrapper = styled.div<CompositionProps>`
-  ${applyStyles};
-  display: ${({ inline }) => (inline ? 'inline-grid' : 'grid')};
-`
-
-const Composition: React.FunctionComponent<CompositionProps> = ({
-  children,
-  ...restProps
-}) => {
+const Composition: React.FunctionComponent<CompositionProps> = (props) => {
+  const { {
+    children,
+    className,
+    inline,
+    ...restProps
+  } = props
   const areaComponents = parseTemplates(restProps)
   const hasAreaComponents = Object.keys(areaComponents).length > 0
   const childrenType = typeof children
@@ -40,11 +39,18 @@ const Composition: React.FunctionComponent<CompositionProps> = ({
   )
 
   return (
-    <CompositionWrapper {...restProps}>
+    <Layout.produceComponent.div
+      className={[
+        className, 
+        Layout.produceStyles(applyStyles(props)),
+        Layout.produceStyles`display: ${() => (inline ? 'inline-grid' : 'grid')};`
+      ].join(' ')}
+      {...restProps}
+    >
       {hasAreaComponents && hasChildrenFunction
         ? (children as ChildrenFunction)(areaComponents)
         : children}
-    </CompositionWrapper>
+    </Layout.produceComponent.div>
   )
 }
 ```

The component is moved inside render method, which is not good for performance. This was done so that the user can configure the library after imports, as it's confusing the end dev has to run `Layout.configure` at a precise point in time (i.e. before importing any components).

**Comment by kettanaito** [comment]
@bitttttten Hi. I will try to put up an specification of the pluggable styling solution in the description above. Please, I would be thankful for any kind of input from your side!

**Comment by bitttttten** [comment]
Yeah, I was thinking about styled-components and how their css api does not handle committing styles the dom like emotion does. [Here you can see my test of exactly this](https://codesandbox.io/s/1qpowv50j4). More work needs to be done for styled components.

And I was wondering about your original idea of having pre-defined plugins, especially for the use case of using a different library. I think it's pretty simple. For example, you could do:

```js
import { Composition } from 'atomic-layout/emotion'
```

Which looks like:

```js
import Layout from './Layout'
import css from 'emotion'

Layout.configure({
	produceStyles: (styles, ...rest) => css([styles], ...rest)
})

export default Layout
export { default as Box } from './components/Box'
export { default as Composition } from './components/Composition'
export { default as Only } from './components/Only'
```

Although this does not address any of other benefits, like decoupling atomic layout from a view library, having atomic layout emit CSS for dev's to handle at build time, and so on. It feels like inspiration should be taken from astroturf for that.

(I am in the middle of a long weekend, so hopefully this comment is clear.. also I will probably come back with more feedback later on! 😛 )

**Comment by kettanaito** [comment]
Hello, @bitttttten. I've established some changes to adopt pluggable styling solution under #145. May I please ask you to checkout that branch and see if it works with `emotion` for you? Your feedback is highly appreciated!

**Comment by kettanaito** [comment]
I've updated the [Pluggable styling API request](https://github.com/kettanaito/atomic-layout/pull/145) with the specification of the feature. It describes the feature behavior and changes introduces by the pull request.

**Comment by hasparus** [comment]
Hey @kettanaito, I just wanted to notice that "SSR with no config" benefit that @bitttttten has mentioned in https://github.com/kettanaito/atomic-layout/issues/122 is only present in `@emotion/core` and `@emotion/styled`.

BTW I just watched _"Creating layouts that last"_. It's a great talk. 

---

> Q: Currently styled returns a React component. It would be nice to abstract from framework-specific return signatures, and operate with emitted class names.

Not every CSS-in-JS framework lets you operate on classNames. `@emotion/core` (Emotion for React) is an example.

---

I wanted to ask, since I experienced that migrations from styled-components to `@emotion/styled` are usually effortless. Does `atomic-layout` use any implementation details of styled-components? I'm wondering about just aliasing `styled-components` to `@emotion/styled` to adopt atomic-layout in projects that already have Emotion.

```
import styled from 'styled-components'
```
turns into
```
import styled from '@emotion/styled'
```

**Comment by kettanaito** [comment]
Hi, @hasparus. Thanks!

Indeed, emotion and styled-components share a similar API, that's why emotion support is one of the first on the table to be adopted. Take a look at [my comment](https://github.com/kettanaito/atomic-layout/issues/245#issuecomment-550442519) in the related issue. It describes where styled-components are used.

TL;DR I think such support should happen, but it needs to be done properly. My main concern is how to build/distribute such variation of a library. For example, if you would like to use it with emotion, which package should be listed in peer dependencies? Both? It looks like it should be distributed as a separate `@atomic-layout/emotion` package to have a valid package.json. Yet, I'm open to brainstorming and suggestions!

**Comment by hasparus** [comment]
wow! Thanks for quick response and for the link to your previous comment.

I think doing similar thing as `rebass` would be pretty great if it worked.
- https://github.com/rebassjs/rebass/blob/master/packages/reflexbox/babel.config.js
- https://rebassjs.org/migrating/

>  For example, if you would like to use it with emotion, which package should be listed in peer dependencies

I guess people would be fine without @emotion/core in peerDependencies. My reasoning behind it is: If someone is looking for a way to use atomic-layout with Emotion, they probably have Emotion already installed.

Swapping import name is a pretty brute way to go, and although it may be a quick win which allows allows introducing atomic-layout to [1.6m weekly downloads](https://www.npmjs.com/package/@emotion/core) of `@emotion/core` userbase adopt atomic-layout, it probably isn't not be the proper abstraction you are looking for.

**Comment by hasparus** [comment]
> I think that with a proper abstraction we can support about any styling solution

Supporting _styled-components-like_ styling solutions by injection of `styled` function seems plausible. It would allow you to cover `styled-components`, `@emotion/styled` and userspace derivatives (building a `styled` function for `theme-ui` wouldn't be too hard, and it would allow to use values from the theme as `gutter` for example).

I can imagine that other styling solutions (well, mby except styledjsx or linaria) could be supported with "support/plugin packages".

**Comment by kettanaito** [comment]
That's a good reference from rebass, thanks! I reckon import replacement can be done via Rollup directly, but since Atomic Layout already uses Babel, it should be fine as well. The only question I have: where should this replacement happen? The end consumer cannot affect the bundled code, so I suppose it must be done during the library's build. Do you have an idea how the build pipeline would look like in this case?

> I think such replacement must live close to the code base and be officially distributed. For example, I'd like to run end-to-end tests on atomic-layout+emotion/styled build to make sure it works per spec. Can we leverage certain CLI flag during installation and `preinstall` npm script to replace the imports?

I would rather define which styling API Atomic Layout expects, than which libraries it supports. Since now we operate on styled-components-like API, I would stick with it for time being. Which makes styling support narrow down to the following solutions, initially:

- styled-components
- emotion/styled
- arbitrary library with `(css: string) => any` signature

**Comment by kettanaito** [comment]
The more I think about the monorepo pattern the more it makes sense to be as a library's developer. When each variation of the library is a separate scoped package it can target variantion-specific implementation, dependencies, and test suites without creating a mess. As mentioned previously, the core of the library can be separated into `atomic-layout/core` quite easily, which is a great monorepo pre-reqiusite.

As the library's consumer, however, I feel struggle, as each library's vairant is a _separate package_. There has to be some naming convetion, like:

- `@atomic-layout/emotion`
- `@atomic-layout/vue`

Such imports are lengthy, and combined with the library's usage rate in a code base can become irritating quite fast. One can solve that with import resolve alias (in case of webpack), but that's not beginner-friendly in general.

**Comment by bitttttten** [comment]
`@atomic-layout/emotion` and `@atomic-layout/vue` look good to me! It also scales to native too with `@atomic-layout/native`.

FYI emotion follows this pattern:

```js
import styled, { css } from '@emotion/native'

const Container = styled.View`
  display: flex;
  justify-content: center;
  align-items: center;
  margin: 50px;
`
```

https://emotion.sh/docs/@emotion/native
https://emotion.sh/docs/@emotion/primitives

I have never used it myself, but the library also exposes a `createEmotion` function: https://emotion.sh/docs/create-emotion. Might be nice for some inspiration.

**Comment by kettanaito** [comment]
@bitttttten thanks for the feedback. The first step to support emotion is coming to an end, I'm yet to verify a few nuances and about to merge. The second step would be to create and publish the actual `@atomic-layout/emotion` package.

**Comment by kettanaito** [comment]
## Official Emotion support

Atomic Layout now distributes a dedicated package that supports Emotion, called [`@atomic-layout/emotion`](https://github.com/kettanaito/atomic-layout/tree/master/packages/atomic-layout-emotion).

```bash
$ npm install @atomic-layout/emotion@0.12.0 @emotion/core @emotion/styled
```

**Comment by bitttttten** [comment]
Ahh you pulled it off! Congrats ✨
